src/Workspaces/Core/Portable/Execution/ChecksumTreeCollection.ChecksumObjectCache.cs

```diff
@@ -0,0 +1,136 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Execution
+{
+    /// <summary>
+    /// This is collection of root checksum tree node
+    /// </summary>
+    internal partial class ChecksumTreeCollection
+    {
+        /// <summary>
+        /// map kind to actual checksum object collection
+        /// 
+        /// since our hierarchical checksum tree should only hold onto green node. 
+        /// different kind of checksum object might share same green node such as document state
+        /// for SourceText and DocumentInfo. so we have this cache type
+        /// </summary>
+        private class ChecksumObjectCache
+        {
+            private ChecksumObject _checksumObject;
+
+            private SubTreeNode _lazyChecksumTree;
+            private ConcurrentDictionary<string, ChecksumObject> _lazyKindToChecksumObjectMap;
```

Comment:
concurrent dictionary is more convenient than lock + dictionary. are you saying concurrent dictionary is less expansive than normal dictionary? otherwise, lock + dictionary seems more work than just using concurrent dictionary.